pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +100,255 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+			continue
+		}
+		if machine.DeletionTimestamp != nil {
+			// Ignore this node since its member previously was, or is in the process of being removed
+			klog.V(4).Infof("Ignoring node (%s) for scale-up since its machine (%s) is pending deletion", node.Name, machine.Name)
+			continue
+		}
+
+		// Wait until the machine has a deletion hook present before adding it as a member
+		if !ceohelpers.HasMachineDeletionHook(machine) {
+			continue
+		}
+
+		// Wait until the etcd pod is running but not ready on the node which indicates
+		// it is waiting to be added to the cluster membership
+		// Note: Checking the pod before adding the member ensures no two unstarted members are added in parallel.
+		// If multiple new nodes are added as members in parallel they will fail with
+		// a discovery error ("error validating peerURLs: ... member count is unequal") because
+		// the unstarted members won't include each other in their ETCD_INITIAL_CLUSTER env.
+		podName := fmt.Sprintf("etcd-%v", node.Name)
+		runningNotReady, err := c.isEtcdContainerRunningNotReady(podName)
+		if err != nil {
+			return "", fmt.Errorf("failed to check if pod is running and not ready: %v", err)
+		}
+		if !runningNotReady {
+			continue
+		}
+
+		// We have a valid new node with a running etcd pod to add as a new member
+		return peerURL, nil
+	}
+	return "", nil
+}
+
+// ensureEtcdLearnerPromotion checks for all learner members in the cluster membership and tries
+// to promote them if they have the required machine deletion hooks present
+func (c *ClusterMemberController) ensureEtcdLearnerPromotion(ctx context.Context, recorder events.Recorder) error {
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+		return fmt.Errorf("could not get etcd member list: %v", err)
+	}
+
+	for _, member := range members {
+		// Check if the member should be promoted
+		promote, err := c.shouldPromote(member)
+		if err != nil {
+			return fmt.Errorf("failed to decide on promotion for member (%s): %v", member.Name, err)
+		}
+		if !promote {
+			continue
+		}
+
+		// Attempt promoting a learner member
+		err = c.etcdClient.MemberPromote(ctx, member)
+		if err != nil {
+			// Note: Cannot use errors.Is(err, etcdserver.ErrLearnerNotReady) as that is always false in this case.
+			// because etcdserver.ErrLearnerNotReady always returns an errors.New() error type
+			// So we compare the error strings here instead
+			if err.Error() == etcdserver.ErrLearnerNotReady.Error() {
+				// Not being ready for promotion is an expected state until the learner catches up
+				klog.V(2).Infof("Not ready for promotion: etcd learner member (%s) is not yet in sync with leader's log ", member.PeerURLs[0])
+				continue
+			}
+			return fmt.Errorf("failed to promote learner member (%s): %v", member.PeerURLs[0], err)
```

Comment:
why do we have to stop here? could we store the err and move on to the next member?